gtsam/nonlinear/FunctorizedFactor.h

```diff
@@ -0,0 +1,123 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ *  @file  FunctorizedFactor.h
+ *  @author Varun Agrawal
+ **/
+
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+#include <cmath>
+
+namespace gtsam {
+
+/**
+ * Factor which evaluates functor and uses the result to compute
+ * error on provided measurement.
+ * The provided FUNCTOR should provide two definitions: `argument_type` which
+ * corresponds to the type of input it accepts and `return_type` which indicates
+ * the type of the return value. This factor uses those type values to construct
+ * the functor.
+ *
+ * Template parameters are
+ * @param FUNCTOR: A class which operates as a functor.
+ */
+template <typename FUNCTOR>
+class GTSAM_EXPORT FunctorizedFactor
+    : public NoiseModelFactor1<typename FUNCTOR::argument_type> {
```

Comment:
Is it possible to allow this to accept more variables?  It seems like it should be reasonably common to have

|h(x1, x2, ...) - z|

where we define a functor for h(x1, x2, ...).

Reply:
Yup it takes any number of arguments. That's why you have the `template <typename... Args>` on the constructor.

Reply:
Sorry to clarify, I meant rather than NoiseModelFactor1, also be able to have NoiseModelFactor2, 3, etc

So right now from the constructor templating it's able to do something like

h(x, theta1, theta2, ...)

where theta are fixed parameters, but I was wondering about multiple variables

Reply:
Ah I see. Yeah that's phase 2. Now that we have this factor working, it should be straightforward to extend it to multiple keys.